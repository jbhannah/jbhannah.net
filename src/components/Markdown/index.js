import PropTypes from "prop-types"
import React, { Fragment } from "react"
import rehypeReact from "rehype-react"
import Blockquote from "../Blockquote"
import { HeadingFactory } from "../Heading"
import Link from "../Link"

const renderAst = new rehypeReact({
  createElement: React.createElement,
  components: {
    a: Link,
    blockquote: Blockquote,
    h2: HeadingFactory("h2"),
    h3: HeadingFactory("h3"),
    h4: HeadingFactory("h4"),
    h5: HeadingFactory("h5"),
    h6: HeadingFactory("h6"),
  },
}).Compiler

export const Markdown = ({ htmlAst, root = Fragment }) => {
  const html = renderAst(htmlAst)

  if (html.type !== root.type) {
    const R = root
    return <R>{html.props.children}</R>
  }

  return html
}

export default Markdown

Markdown.propTypes = {
  htmlAst: PropTypes.object,
  root: PropTypes.node,
}
