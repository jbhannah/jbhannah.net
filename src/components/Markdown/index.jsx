import PropTypes from "prop-types"
import React, { Fragment } from "react"
import rehypeReact from "rehype-react"
import visit from "unist-util-visit"
import Blockquote from "../Blockquote"
import Code from "../Code"
import { HeadingFactory } from "../Heading"
import Link from "../Link"

const headingPattern = /^h(\d)$/

const isTag = (node, tagName) =>
  node.hasOwnProperty("type") &&
  node.type === "element" &&
  node.hasOwnProperty("tagName") &&
  node.tagName === tagName

const newSection = children => {
  const section = { type: "element", tagName: "section" }
  section.children = children(section)
  return section
}

const renderAst = htmlAst => {
  let currentLevel, currentParent, currentSection

  visit(htmlAst, (node, index, parent) => {
    if (node.type === "root") return visit.CONTINUE

    if (node.type === "element" && headingPattern.test(node.tagName)) {
      const level = node.tagName.match(headingPattern)[1]

      if (level > currentLevel) {
        const section = newSection(() => parent.children.splice(index, 1))
        currentSection.children.push(section)
        currentSection = section
      } else {
        currentSection = newSection(section =>
          parent.children.splice(index, 1, section)
        )
      }

      currentLevel = level
      return visit.SKIP
    }

    const prevSibling = parent.children[index - 1]

    if (
      typeof currentSection === "undefined" ||
      typeof prevSibling === "undefined" ||
      parent === currentParent ||
      (isTag(node, "div") &&
        node.properties.hasOwnProperty("class") &&
        node.properties.class.includes("footnotes"))
    ) {
      return visit.SKIP
    }

    currentParent = node

    if (isTag(prevSibling, "section")) {
      const [curr] = parent.children.splice(index, 1)
      currentSection.children.push(curr)
      return index
    }

    return index - 1
  })

  return new rehypeReact({
    createElement: React.createElement,
    components: {
      a: Link,
      blockquote: Blockquote,
      code: Code,
      h2: HeadingFactory("h2"),
      h3: HeadingFactory("h3"),
      h4: HeadingFactory("h4"),
      h5: HeadingFactory("h5"),
      h6: HeadingFactory("h6"),
    },
  }).Compiler(htmlAst)
}

export const Markdown = ({ htmlAst, root = Fragment }) => {
  const html = renderAst(htmlAst)
  const R = root

  return <R {...html.props} />
}

export default Markdown

Markdown.propTypes = {
  htmlAst: PropTypes.object,
  root: PropTypes.node,
}
