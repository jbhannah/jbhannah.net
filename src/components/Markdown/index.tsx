import { Comment, Element, Node, Root, Text } from "hast"
import * as React from "react"
import rehypeReact from "rehype-react"
import visit from "unist-util-visit"
import { Blockquote } from "../Blockquote"
import { Code } from "../Code"
import { H2, H3, H4, H5, H6 } from "../Heading"
import { Link } from "../Link"

const headingPattern = /^h(\d)$/

const isTag = (node: Node, tagName: string) =>
  node.type === "element" && (node as Element).tagName === tagName

const newSection = (
  children: (section: Element) => (Element | Comment | Text)[]
) => {
  const section: Element = { type: "element", tagName: "section", children: [] }
  section.children = children(section)
  return section
}

const compiler = new rehypeReact({
  createElement: React.createElement,
  components: {
    a: Link,
    blockquote: Blockquote,
    code: Code,
    h2: H2,
    h3: H3,
    h4: H4,
    h5: H5,
    h6: H6,
  },
}).Compiler

const renderAst = (htmlAst: Node) => {
  let currentLevel: number, currentParent: Element, currentSection: Element

  visit(htmlAst, (node: Element | Root, index, parent: Element) => {
    if (node.type === "root") return visit.CONTINUE

    if (
      node.type === "element" &&
      headingPattern.test((node as Element).tagName)
    ) {
      const level: number = parseInt(
        (node as Element).tagName.match(headingPattern)[1]
      )

      if (level > currentLevel) {
        const section = newSection(() => parent.children.splice(index, 1))
        currentSection.children.push(section)
        currentSection = section
      } else {
        currentSection = newSection((section) =>
          parent.children.splice(index, 1, section)
        )
      }

      currentLevel = level
      return visit.SKIP
    }

    const prevSibling = parent.children[index - 1]

    if (
      typeof currentSection === "undefined" ||
      typeof prevSibling === "undefined" ||
      parent === currentParent ||
      (isTag(node, "div") &&
        node.properties.hasOwnProperty("class") &&
        (node.properties.class as string).includes("footnotes"))
    ) {
      return visit.SKIP
    }

    currentParent = node

    if (isTag(prevSibling, "section")) {
      const [curr] = parent.children.splice(index, 1)
      currentSection.children.push(curr)
      return index
    }

    return index - 1
  })

  return compiler(htmlAst)
}

interface MarkdownProps {
  htmlAst: Node
  root?: React.ReactType
}

export const Markdown: React.FunctionComponent<MarkdownProps> = ({
  htmlAst,
  root: R = React.Fragment,
}) => {
  const html = renderAst(htmlAst)

  return <R {...html.props} />
}
