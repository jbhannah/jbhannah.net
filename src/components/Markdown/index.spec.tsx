import { mount } from "enzyme"
import * as React from "react"
import Markdown from "."

const htmlAst = {
  type: "root",
  children: [
    {
      type: "element",
      tagName: "p",
      properties: {},
      children: [
        {
          type: "text",
          value: "Test paragraph.",
        },
      ],
    },
    {
      type: "element",
      tagName: "h2",
      properties: {},
      children: [
        {
          type: "text",
          value: "Test heading",
        },
      ],
    },
    {
      type: "element",
      tagName: "blockquote",
      properties: {},
      children: [
        {
          type: "element",
          tagName: "p",
          children: [
            {
              type: "text",
              value: "Test blockquote",
            },
          ],
        },
      ],
    },
    {
      type: "element",
      tagName: "h3",
      properties: {},
      children: [
        {
          type: "text",
          value: "Test subheading",
        },
      ],
    },
    {
      type: "element",
      tagName: "a",
      properties: {
        href: "https://www.example.com",
      },
      children: [
        {
          type: "text",
          value: "Test link",
        },
      ],
    },
    {
      type: "element",
      tagName: "p",
      properties: {},
      children: [
        {
          type: "text",
          value: "Test paragraph 2.",
        },
      ],
    },
    {
      type: "element",
      tagName: "h2",
      properties: {},
      children: [
        {
          type: "text",
          value: "Test heading 2",
        },
      ],
    },
    {
      type: "element",
      tagName: "p",
      properties: {},
      children: [
        {
          type: "text",
          value: "Test paragraph 3.",
        },
      ],
    },
    {
      type: "element",
      tagName: "div",
      properties: {
        class: "footnotes",
      },
      children: [
        {
          type: "text",
          value: "Test footnotes.",
        },
      ],
    },
  ],
  data: {
    quirksMode: false,
  },
}

const TestMarkdown = (props) => mount(<Markdown {...props} />)

describe("Markdown", () => {
  describe("without a root element", () => {
    const tree = TestMarkdown({ htmlAst })
    const headings = tree.find("Heading")

    it("renders correctly", () => {
      expect(tree).toMatchSnapshot()
    })

    it("renders <h#> tags as <Heading> components", () => {
      expect(headings).toHaveLength(3)
      expect(headings.find("h2")).toHaveLength(2)
      expect(headings.find("h3")).toHaveLength(1)
    })

    it("renders <a> tags as <Link> components", () => {
      expect(tree.find("Link")).toHaveLength(1)
    })

    it("wraps contents after headings in <section> tags", () => {
      expect(tree.children("section")).toHaveLength(2)
      expect(tree.find("section")).toHaveLength(3)
    })
  })

  describe("with a root element", () => {
    const tree = TestMarkdown({ htmlAst, root: "article" })

    it("renders correctly", () => {
      expect(tree).toMatchSnapshot()
    })

    it("wraps the tree in the specified element", () => {
      expect(tree.children("article")).toHaveLength(1)
    })
  })
})
