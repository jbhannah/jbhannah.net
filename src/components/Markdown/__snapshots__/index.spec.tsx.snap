// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown with a root element renders correctly 1`] = `
<Markdown
  htmlAst={
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Test paragraph.",
            },
          ],
          "properties": Object {},
          "tagName": "p",
          "type": "element",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Test heading",
                },
              ],
              "properties": Object {},
              "tagName": "h2",
              "type": "element",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "Test blockquote",
                    },
                  ],
                  "tagName": "p",
                  "type": "element",
                },
              ],
              "properties": Object {},
              "tagName": "blockquote",
              "type": "element",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "Test subheading",
                    },
                  ],
                  "properties": Object {},
                  "tagName": "h3",
                  "type": "element",
                },
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "Test link",
                    },
                  ],
                  "properties": Object {
                    "href": "https://www.example.com",
                  },
                  "tagName": "a",
                  "type": "element",
                },
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "Test paragraph 2.",
                    },
                  ],
                  "properties": Object {},
                  "tagName": "p",
                  "type": "element",
                },
              ],
              "tagName": "section",
              "type": "element",
            },
          ],
          "tagName": "section",
          "type": "element",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Test heading 2",
                },
              ],
              "properties": Object {},
              "tagName": "h2",
              "type": "element",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Test paragraph 3.",
                },
              ],
              "properties": Object {},
              "tagName": "p",
              "type": "element",
            },
          ],
          "tagName": "section",
          "type": "element",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Test footnotes.",
            },
          ],
          "properties": Object {
            "class": "footnotes",
          },
          "tagName": "div",
          "type": "element",
        },
      ],
      "data": Object {
        "quirksMode": false,
      },
      "type": "root",
    }
  }
  root="article"
>
  <article>
    <p
      key="h-2"
    >
      Test paragraph.
    </p>
    <section
      key="h-3"
    >
      <Component
        key="h-4"
      >
        <Heading
          level="h2"
        >
          <Styled(h2)>
            <h2
              className="css-1qf3p84"
            >
              Test heading
            </h2>
          </Styled(h2)>
        </Heading>
      </Component>
      <Blockquote
        key="h-5"
      >
        <Styled(blockquote)>
          <blockquote
            className="css-2570gk"
          >
            <p
              key="h-6"
            >
              Test blockquote
            </p>
          </blockquote>
        </Styled(blockquote)>
      </Blockquote>
      <section
        key="h-7"
      >
        <Component
          key="h-8"
        >
          <Heading
            level="h3"
          >
            <Styled(h3)>
              <h3
                className="css-1qf3p84"
              >
                Test subheading
              </h3>
            </Styled(h3)>
          </Heading>
        </Component>
        <Link
          href="https://www.example.com"
          key="h-9"
        >
          <Styled(a)
            href="https://www.example.com"
            onClick={[Function]}
            rel="noopener"
            target="_blank"
          >
            <a
              className="css-nk3r0j"
              href="https://www.example.com"
              onClick={[Function]}
              rel="noopener"
              target="_blank"
            >
              Test link
            </a>
          </Styled(a)>
        </Link>
        <p
          key="h-10"
        >
          Test paragraph 2.
        </p>
      </section>
    </section>
    <section
      key="h-11"
    >
      <Component
        key="h-12"
      >
        <Heading
          level="h2"
        >
          <Styled(h2)>
            <h2
              className="css-1qf3p84"
            >
              Test heading 2
            </h2>
          </Styled(h2)>
        </Heading>
      </Component>
      <p
        key="h-13"
      >
        Test paragraph 3.
      </p>
    </section>
    <div
      className="footnotes"
      key="h-14"
    >
      Test footnotes.
    </div>
  </article>
</Markdown>
`;

exports[`Markdown without a root element renders correctly 1`] = `
<Markdown
  htmlAst={
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Test paragraph.",
            },
          ],
          "properties": Object {},
          "tagName": "p",
          "type": "element",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Test heading",
                },
              ],
              "properties": Object {},
              "tagName": "h2",
              "type": "element",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "Test blockquote",
                    },
                  ],
                  "tagName": "p",
                  "type": "element",
                },
              ],
              "properties": Object {},
              "tagName": "blockquote",
              "type": "element",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "Test subheading",
                    },
                  ],
                  "properties": Object {},
                  "tagName": "h3",
                  "type": "element",
                },
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "Test link",
                    },
                  ],
                  "properties": Object {
                    "href": "https://www.example.com",
                  },
                  "tagName": "a",
                  "type": "element",
                },
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "Test paragraph 2.",
                    },
                  ],
                  "properties": Object {},
                  "tagName": "p",
                  "type": "element",
                },
              ],
              "tagName": "section",
              "type": "element",
            },
          ],
          "tagName": "section",
          "type": "element",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Test heading 2",
                },
              ],
              "properties": Object {},
              "tagName": "h2",
              "type": "element",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Test paragraph 3.",
                },
              ],
              "properties": Object {},
              "tagName": "p",
              "type": "element",
            },
          ],
          "tagName": "section",
          "type": "element",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Test footnotes.",
            },
          ],
          "properties": Object {
            "class": "footnotes",
          },
          "tagName": "div",
          "type": "element",
        },
      ],
      "data": Object {
        "quirksMode": false,
      },
      "type": "root",
    }
  }
>
  <p
    key="h-2"
  >
    Test paragraph.
  </p>
  <section
    key="h-3"
  >
    <Component
      key="h-4"
    >
      <Heading
        level="h2"
      >
        <Styled(h2)>
          <h2
            className="css-1qf3p84"
          >
            Test heading
          </h2>
        </Styled(h2)>
      </Heading>
    </Component>
    <Blockquote
      key="h-5"
    >
      <Styled(blockquote)>
        <blockquote
          className="css-2570gk"
        >
          <p
            key="h-6"
          >
            Test blockquote
          </p>
        </blockquote>
      </Styled(blockquote)>
    </Blockquote>
    <section
      key="h-7"
    >
      <Component
        key="h-8"
      >
        <Heading
          level="h3"
        >
          <Styled(h3)>
            <h3
              className="css-1qf3p84"
            >
              Test subheading
            </h3>
          </Styled(h3)>
        </Heading>
      </Component>
      <Link
        href="https://www.example.com"
        key="h-9"
      >
        <Styled(a)
          href="https://www.example.com"
          onClick={[Function]}
          rel="noopener"
          target="_blank"
        >
          <a
            className="css-nk3r0j"
            href="https://www.example.com"
            onClick={[Function]}
            rel="noopener"
            target="_blank"
          >
            Test link
          </a>
        </Styled(a)>
      </Link>
      <p
        key="h-10"
      >
        Test paragraph 2.
      </p>
    </section>
  </section>
  <section
    key="h-11"
  >
    <Component
      key="h-12"
    >
      <Heading
        level="h2"
      >
        <Styled(h2)>
          <h2
            className="css-1qf3p84"
          >
            Test heading 2
          </h2>
        </Styled(h2)>
      </Heading>
    </Component>
    <p
      key="h-13"
    >
      Test paragraph 3.
    </p>
  </section>
  <div
    className="footnotes"
    key="h-14"
  >
    Test footnotes.
  </div>
</Markdown>
`;
